stdlib/stdlib-external/string.md

```diff
@@ -0,0 +1,56 @@
+# Grain strings
+
+Grain uses Unicode, specifically UTF-8 for strings.
+
+## Definitions
+
+- _code point_ an atomic unit of information. Text is a sequence of code points.
+- _code unit_ a unit of storage of an encoded code point.
+- _grapheme_ a sequence of one or more code points displayed as a single
+  graphical element.
+
+## Encoding
+
+Unicode is the universal character encoding. Each character has a code point.
+Valid code points are in the ranges 0x0000...0xD7FF and 0xE000...0x10FFFF.
+The gap between the ranges is for surrogate pairs.
+
+Several schemes exist to encode Unicode. Grain uses UTF-8. The 8 indicates the
+number of bits in a code unit. UTF-8 doesn't use surrogate pairs so you don't
+have to worry about that gap between the ranges.
+
+Code points are encoded in the following way. A single code point can be up to
+four bytes.
+
+```
+    0xxxxxxx
+    110xxxxx 10xxxxxx
+    1110xxxx 10xxxxxx 10xxxxxx
+    11110xxx 10xxxxxx 10xxxxxx 10xxxxxx
+```
+
+In order to know how many code points a string contains you count the leading
+bytes which begin with either 00, 01 or 11. This is given by the following
+expressions in AssemblyScript.
+
+```
+    (byte & 0b11000000) !== 0b01000000
+    or
+    (byte & 0xC0) !== 0x80)
+```
+
+## Combining and normalization
+
+Some graphemes are encoded by multiple code points. Some graphemes can be
+encoded by either a single code point or multiple code points. For example,
+__ñ__ is U+00F1, but it can also be encoded by combining __n__ U+0006E and tilde
+ __◌̃__ U+0303.
+
+ Normalization is the process by which combining code points are reduced to
+ either single code points or put into canonical order. This is necessary for
+ searching and comparing strings. Strings in Grain are not normalized.
+
+ ## JavaScript
+
+ JavaScript strings are not encoded in UTF-8.
+ 
```

Comment:
I don't think we need to mention this.
```suggestion
```

Reply:
I think this is important to mention in the contributor docs, especially as we expand our JS FFI. Though I think we should expand on it and maybe point out https://github.com/grain-lang/grain/blob/master/runtime/src/utils/utils.js#L76 specifically.

Reply:
We certainly can mention it in the contributor docs, but I should note that I wouldn't expect it to impact many users, since the Grain FFI layer translates directly from our UTF-8 strings into native ones with whatever encoding JS uses (I forget off the top of my head).

Reply:
JS uses UCS-2 with unvalidated surrogate pairs. A DOMString is the same but with a replacement character for invalid surrogate pairs. A USVString is the same but with validated surrogate pairs so it's UTF-16 restricted to the Basic Multilingual Plane. A CSSOMString is the same as a DOMString. The encoder converts to USVString.

Reply:
@belph These are contributor docs. I agree that we shouldn't mention this on the website or anything.

@JohnnyRidout Could you add that to this section?

Reply:
@ospencer,   I've added it and moved the document to docs/contributor.

I can add `\u0000` style escapes in another pull request next weekend.

@belph I can update the user docs with information about the escape sequences next weekend.

Does that cover everything?

Reply:
Yes, this is awesome! Thanks for your hard work on this!!